# Data Cleaning and EDA Prep Notebook

In [1]:
# Import the required libraries
import pandas as pd
import numpy as np

# set up pandas to display floats in a more human friendly way
pd.options.display.float_format = '{:,.2f}'.format

### Original training data files downloaded from data source were split between values and labels. 
#### Step 1: Conduct data cleaning USING FUNCTIONS on the Training and Test Files
#### Step 2: Combine the Training values and labels into a single file for use in model training.

In [2]:
# TRAINING DATA
train_values_raw = pd.read_csv('../data/original/TrainingSetValues/4910797b-ee55-40a7-8668-10efd5c1b960.csv')

print(train_values_raw.shape)
train_values_raw.head(3)

(59400, 40)


id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572    6,000.00    2011-03-14         Roman        1390         Roman   
1   8776        0.00    2013-03-06       Grumeti        1399       GRUMETI   
2  34310       25.00    2013-02-25  Lottery Club         686  World vision   

   longitude  latitude     wpt_name  num_private  ... payment_type  \
0      34.94     -9.86         none            0  ...     annually   
1      34.70     -2.15     Zahanati            0  ...    never pay   
2      37.46     -3.82  Kwa Mahundi            0  ...   per bucket   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  

[3 rows x 40 columns]

In [4]:
# TRAIN TARGET
train_targets = pd.read_csv('../data/original/TrainingSetLabels/0bf8bc6e-30d0-4c50-956a-603fc693d966.csv')

print(train_targets.shape)
train_targets.head(3)

(59400, 2)


id status_group
0  69572   functional
1   8776   functional
2  34310   functional

In [5]:
# VALIDATION DATA 
test_values_raw = pd.read_csv('../data/original/TestSetValues/702ddfc5-68cd-4d1d-a0de-f5f566f76d91.csv')

print(test_values_raw.shape)
test_values_raw.head(3)

(14850, 40)


id  amount_tsh date_recorded                  funder  gps_height  \
0  50785        0.00    2013-02-04                    Dmdd        1996   
1  51630        0.00    2013-02-04  Government Of Tanzania        1569   
2  17168        0.00    2013-02-01                     NaN        1567   

  installer  longitude  latitude                 wpt_name  num_private  ...  \
0      DMDD      35.29     -4.06  Dinamu Secondary School            0  ...   
1       DWE      36.66     -3.31                  Kimnyak            0  ...   
2       NaN      34.77     -5.00           Puma Secondary            0  ...   

  payment_type water_quality quality_group      quantity  quantity_group  \
0    never pay          soft          good      seasonal        seasonal   
1    never pay          soft          good  insufficient    insufficient   
2    never pay          soft          good  insufficient    insufficient   

                 source           source_type  source_class  \
0  rainwater harvesting  rainwater harvesting       surface   
1                spring                spring   groundwater   
2  rainwater harvesting  rainwater harvesting       surface   

      waterpoint_type waterpoint_type_group  
0               other                 other  
1  communal standpipe    communal standpipe  
2               other                 other  

[3 rows x 40 columns]

## Column and Row Information
* 40 columns/features in the raw data TRAINING and TEST
* 59,400 rows in the TRAINING data
    * TRAINING labels: 
        * functional                 32,259
        * non functional             22,824
        * functional needs repair     4,317

* 14,850 rows in the TEST data

In [6]:
train_values_raw.dtypes

id                         int64
amount_tsh               float64
date_recorded             object
funder                    object
gps_height                 int64
installer                 object
longitude                float64
latitude                 float64
wpt_name                  object
num_private                int64
basin                     object
subvillage                object
region                    object
region_code                int64
district_code              int64
lga                       object
ward                      object
population                 int64
public_meeting            object
recorded_by               object
scheme_management         object
scheme_name               object
permit                    object
construction_year          int64
extraction_type           object
extraction_type_group     object
extraction_type_class     object
management                object
management_group          object
payment                   object
payment_ty

In [7]:
for var in train_values_raw.columns:
    # print the first 20 unique values in the cols
    unique_vals = train_values_raw[var].unique()
    print(var, unique_vals.size, unique_vals[0:20], '\n')

id 59400 [69572  8776 34310 67743 19728  9944 19816 54551 53934 46144 49056 50409
 36957 50495 53752 61848 48451 58155 34169 18274] 

amount_tsh 98 [6000.    0.   25.   20.  200.  500.   50. 4000. 1500.    6.  250.   10.
 1000.  100.   30. 2000.  400. 1200.   40.  300.] 

date_recorded 356 ['2011-03-14' '2013-03-06' '2013-02-25' '2013-01-28' '2011-07-13'
 '2011-03-13' '2012-10-01' '2012-10-09' '2012-11-03' '2011-08-03'
 '2011-02-20' '2013-02-18' '2012-10-14' '2013-03-15' '2012-10-20'
 '2011-08-04' '2011-07-04' '2011-09-04' '2011-07-22' '2011-02-22'] 

funder 1898 ['Roman' 'Grumeti' 'Lottery Club' 'Unicef' 'Action In A'
 'Mkinga Distric Coun' 'Dwsp' 'Rwssp' 'Wateraid' 'Isingiro Ho' 'Private'
 'Danida' 'World Vision' 'Lawatefuka Water Supply' 'Biore' 'Rudep'
 'Hesawa' 'Twe' 'Isf' 'African Development Bank'] 

gps_height 2428 [1390 1399  686  263    0   62 1062 1368 1645 1703 1656 1162 1763 2216
 1510  672 1273  200 1443 1256] 

installer 2146 ['Roman' 'GRUMETI' 'World vision' 'UNICEF' 'A

## Original Data Column Descriptions

* id - Numeric identifer for the waterpoint
* amount_tsh - Total static head (amount water available to waterpoint)
* date_recorded - The date the row was entered
* funder - Who funded the well
* gps_height - Altitude of the well
* installer - Organization that installed the well
* longitude - GPS coordinate
* latitude - GPS coordinate
* wpt_name - Name of the waterpoint if there is one
* num_private -
* basin - Geographic water basin
* subvillage - Geographic location
* region - Geographic location, NOTE: Hierarchy is Region > LGA > Ward
* region_code - Geographic location (coded)
* district_code - Geographic location (coded)
* lga - Geographic location
* ward - Geographic location
* population - Population around the well
* public_meeting - True/False
* recorded_by - Group entering this row of data
* scheme_management - Who operates the waterpoint
* scheme_name - Who operates the waterpoint
* permit - True/False, If the waterpoint is permitted
* construction_year - Year the waterpoint was constructed
* extraction_type - The kind of extraction the waterpoint uses
* extraction_type_group - The kind of extraction the waterpoint uses
* extraction_type_class - The kind of extraction the waterpoint uses
* management - How the waterpoint is managed
* management_group - How the waterpoint is managed
* payment - What the water costs
* payment_type - What the water costs
* water_quality - The quality of the water
* quality_group - The quality of the water
* quantity - The quantity of water
* quantity_group - The quantity of water
* source - The source of the water
* source_type - The source of the water
* source_class - The source of the water
* waterpoint_type - The kind of waterpoint
* waterpoint_type_group - The kind of waterpoint



## Data cleaning steps:
* __Duplicate check__ 

* __Address Null Values__

* __Address Zeros in Numeric Values__

* __String type normalization__

* __Note:__ No Data type conversions needed

### Duplicate Check
Spoiler Alert: There are no exact duplicate rows nor duplicate identifers in the Training or Test dataset

In [8]:
# Functions for Duplicate checks 
def has_exact_dups(df):
    dups = df[df.duplicated()]
    return len(dups) > 0

def has_identifier_dups(df, col_name='id'):
    num_rows = df.shape[0]
    num_ids = len(df[col_name].unique())
    return num_ids != num_rows

In [9]:
# Dup checking
dup1 = has_exact_dups(train_values_raw)
dup2 = has_identifier_dups(train_values_raw)
dup3 = has_exact_dups(test_values_raw)
dup4 = has_identifier_dups(test_values_raw)
print(dup1, dup2, dup3, dup4)

False False False False


In [10]:
# Make a deep copy before any data cleaning (Deep copy has own copy of data and index)
train_values_processed = train_values_raw.copy(deep=True)
test_values_processed = test_values_raw.copy(deep=True)

### Handling Null Values
#### TRAINING and TEST columns with Null values:
* funder : set to 'unknown' when null
* installer : set to 'unknown' when null
* subvillage : set to 'unknown' when null
* public_meeting : set to True when null
    * Training dataset: only 5055 out of 59,400 were False. 51,011 out of 59,4000 were True. 
    * Test dataset: only 1291 out of 14,850 were False, 12,738 out of 14,850 were True.
    * As the vast majority of both Training and Test waterpoints have public_meeting of True, use True to replace all nulls.
* scheme_management : set to 'unknown' when null
* scheme_name : set to 'unknown' when null
* permit : set to True when nulls
    * Training dataset: 38,852 out of 59,4000 were True. 17,492 out of 59,4000 where False. 
    * Test dataset:  9754 out of 14,850 were True. 4359 out of 14,850 were False
    * As the majority of both Training and Test waterpoints have permit populated as True, use True to replace all nulls.


In [11]:
# Null handler functions
def handle_nulls_inplace(df, cols_to_fill):
    for item in cols_to_fill:
        for key, value in item.items():
            df[key].fillna(value, inplace=True)

See what percentage of data is missing

In [12]:
train_values_processed.isnull().mean()

id                      0.00
amount_tsh              0.00
date_recorded           0.00
funder                  0.06
gps_height              0.00
installer               0.06
longitude               0.00
latitude                0.00
wpt_name                0.00
num_private             0.00
basin                   0.00
subvillage              0.01
region                  0.00
region_code             0.00
district_code           0.00
lga                     0.00
ward                    0.00
population              0.00
public_meeting          0.06
recorded_by             0.00
scheme_management       0.07
scheme_name             0.47
permit                  0.05
construction_year       0.00
extraction_type         0.00
extraction_type_group   0.00
extraction_type_class   0.00
management              0.00
management_group        0.00
payment                 0.00
payment_type            0.00
water_quality           0.00
quality_group           0.00
quantity                0.00
quantity_group

In [13]:
cols_to_fill = [{'funder': 'unknown'}, {'installer': 'unknown'}, {'subvillage': 'unknown'}, {'public_meeting': True}, {'scheme_management': 'unknown'}, {'scheme_name': 'unknown'}, {'permit': True}]

handle_nulls_inplace(train_values_processed, cols_to_fill)
handle_nulls_inplace(test_values_processed, cols_to_fill)


### Handing Zeros in Numeric Columns
* No change needed
    * latitude : No zeros
    * region_code : No zeros
* __Drop__  data
    * num_private : ~98 of Train,  __DROP this COLUMN__ from Train and Test dataset.

* Replace Zeros
    * construction_year : ~35% of Train and ~35% of Test - Update zeros with the Average Construction Year.

* Do nothing. These 0 values seem in line with data used on the Offical Tanzanian Water Point Mapping System (WPMS)  [http://wpm.maji.go.tz/]. I don't have enough context to know what to replace the zero values with.
    * amount_tsh 
    * gps_height 
    * population 


In [14]:
# Functions for handling Zeros in Numeric columns

def count_zeros(df, col_name):
    return df[df[col_name]==0][col_name].count()


In [15]:
numeric_col_names = ['amount_tsh', 'gps_height', 'latitude', 'longitude', 'num_private', 'region_code', 'district_code', 'population', 'construction_year']

for col_name in numeric_col_names:
    the_train_count = count_zeros(train_values_processed, col_name)
    the_test_count = count_zeros(test_values_processed, col_name)
    if(the_test_count + the_train_count > 0):
        print('TRAIN:', col_name, the_train_count)
        print('TEST:', col_name, the_test_count)

TRAIN: amount_tsh 41639
TEST: amount_tsh 10410
TRAIN: gps_height 20438
TEST: gps_height 5211
TRAIN: longitude 1812
TEST: longitude 457
TRAIN: num_private 58643
TEST: num_private 14656
TRAIN: district_code 23
TEST: district_code 4
TRAIN: population 21381
TEST: population 5453
TRAIN: construction_year 20709
TEST: construction_year 5260


In [16]:
# Drop the num_private COLUMN
train_values_processed.drop('num_private', axis=1, inplace=True)
test_values_processed.drop('num_private', axis=1, inplace=True)

In [17]:

# Drop the rows with 0 longitude
train_indices_long = train_values_processed[train_values_processed['longitude'] == 0 ].index
#test_indices_long = test_values_processed[test_values_processed['longitude'] == 0 ].index
train_values_processed.drop(train_indices_long, inplace=True)
#test_values_processed.drop(test_indices_long, inplace=True)

# drop the rows with 0 district_code
train_indices_distric_code = train_values_processed[train_values_processed['district_code'] == 0 ].index
#test_indices_distric_code = test_values_processed[test_values_processed['district_code'] == 0 ].index
train_values_processed.drop(train_indices_distric_code, inplace=True)
#test_values_processed.drop(test_indices_distric_code, inplace=True)


In [18]:
# Get the average construction year for TRAIN and TEST
known_const_year_rows = train_values_processed[train_values_processed['construction_year']>0]
avg_counstruction_year = int(known_const_year_rows['construction_year'].mean().round())
print(avg_counstruction_year)

test_known_const_year_rows = test_values_processed[test_values_processed['construction_year']>0]
test_avg_counstruction_year = int(test_known_const_year_rows['construction_year'].mean().round())
print(test_avg_counstruction_year)

1997
1997


In [19]:
# Set construction_year to the average construction year where that value is 0
train_values_processed['construction_year'] = train_values_processed.apply(lambda row: avg_counstruction_year if row['construction_year'] == 0 else row['construction_year'], axis=1)
test_values_processed['construction_year'] = test_values_processed.apply(lambda row: test_avg_counstruction_year if row['construction_year'] == 0 else row['construction_year'], axis=1)

### Normalize String columns - all to lower case

* funder - Who funded the well
* installer - Organization that installed the well
* wpt_name - Name of the waterpoint if there is one
* basin - Geographic water basin
* subvillage - Geographic location
* region - Geographic location
* lga - Geographic location
* ward - Geographic location
* recorded_by - Group entering this row of data
* scheme_management - Who operates the waterpoint
* scheme_name - Who operates the waterpoint
* extraction_type - The kind of extraction the waterpoint uses
* extraction_type_group - The kind of extraction the waterpoint uses
* extraction_type_class - The kind of extraction the waterpoint uses
* management - How the waterpoint is managed
* management_group - How the waterpoint is managed
* payment_type - What the water costs
* payment_type - What the water costs
* water_quality - The quality of the water
* quality_group - The quality of the water
* quantity - The quantity of water
* quantity_group - The quantity of water
* source - The source of the water
* source_type - The source of the water
* source_class - The source of the water
* waterpoint_type - The kind of waterpoint
* waterpoint_type_group - The kind of waterpoin


In [20]:
# Normalize String values function(s)
def normalize_strings(df, col_name):
     df[col_name] = df.apply(lambda row: row[col_name].lower(), axis=1)

In [21]:

string_col_names = ['funder', 'installer', 'wpt_name', 'basin', 'subvillage', 'region', 'lga', 'ward', 'recorded_by','scheme_management','scheme_name', 'extraction_type', 'extraction_type_group', 'extraction_type_class', 'management', 'management_group', 'payment', 'payment_type', 'water_quality', 'quality_group', 'quantity', 'quantity_group', 'source', 'source_type' , 'source_class', 'waterpoint_type', 'waterpoint_type_group']

for col_name in string_col_names:
    normalize_strings(train_values_processed, col_name)
    normalize_strings(test_values_processed, col_name)


### New Columns/Feature creation for EDA - TRAINING dataset ONLY
* recorded_year -  Pulling out the year from date_recorded
* waterpoint_age - Calculate as recorded_year - construction_year
* recorded_good_quality - True if quality_group == 'good', False if  anything other than 'good'
* recorded_good_quantity - True if quanity_group == 'sufficient', False if anythign other than 'sufficient'


In [22]:
# Functions for creating new features
def get_recorded_year(recorded_date_string):
    year = 0
    date_segs = recorded_date_string.split('-')
    if((len(date_segs) == 3) & (len(date_segs[0]) == 4)):
        try:
            year = int(date_segs[0])
        except:
            print("Not a valid year format.") 
    return year


def get_waterpoint_age(recorded_year, constructed_year):
    age = 0
    is_logical_year = constructed_year > 0
    is_logical_age = recorded_year > constructed_year
    if (is_logical_year & is_logical_age):
        age = recorded_year - constructed_year
    return age


def get_recorded_good_quality(quality_group):
    result = False
    if ('good' == quality_group):
        result = True
    return result


def get_recorded_good_quanity(quanity_group):
    result = False
    if ('enough' == quanity_group):
        result = True
    return result
        

In [23]:
# recorded_year
train_values_processed['recorded_year'] = train_values_processed.apply(lambda row: get_recorded_year(row['date_recorded']), axis=1)


In [24]:
# waterpoint_age
train_values_processed['waterpoint_age'] = train_values_processed.apply(lambda row: get_waterpoint_age(row['recorded_year'],row['construction_year']), axis=1)


In [25]:
# recorded_good_quality
train_values_processed['recorded_good_quality'] = train_values_processed.apply(lambda row: get_recorded_good_quality(row['quality_group']), axis=1)


In [26]:
# recorded_good_quanity
train_values_processed['recorded_good_quantity'] = train_values_processed.apply(lambda row: get_recorded_good_quanity(row['quantity_group']), axis=1)


## Final Prep
* Add the class labels to the TRAINING dataset
* Save both cleaned TRAINING and TEST to file for use in EDA and Classifer Modeling


In [27]:
train_values_processed_and_labeled = pd.merge(train_values_processed, train_targets, on='id')

In [28]:
train_values_processed_and_labeled.to_csv('../data/train_processed_labeled.csv', index=False)
test_values_processed.to_csv('../data/test_processed.csv', index=False)

In [29]:
train_values_processed_and_labeled.shape

(57565, 44)

In [30]:
test_values_processed.shape

(14850, 39)

In [31]:
train_values_processed.describe()

id  amount_tsh  gps_height  longitude  latitude  region_code  \
count 57,565.00   57,565.00   57,565.00  57,565.00 57,565.00    57,565.00   
mean  37,108.80      327.78      689.60      35.15     -5.89        15.22   
std   21,454.23    3,044.43      693.57       2.61      2.81        17.86   
min        0.00        0.00      -90.00      29.61    -11.65         1.00   
25%   18,528.00        0.00        0.00      33.28     -8.65         5.00   
50%   37,057.00        0.00      427.00      35.00     -5.17        12.00   
75%   55,669.00       30.00    1,332.00      37.23     -3.37        17.00   
max   74,247.00  350,000.00    2,770.00      40.35     -1.00        99.00   

       district_code  population  construction_year  recorded_year  \
count      57,565.00   57,565.00          57,565.00      57,565.00   
mean            5.73      185.64           1,996.88       2,011.92   
std             9.76      477.83              10.23           0.96   
min             1.00        0.00           1,960.00       2,002.00   
25%             2.00        0.00           1,995.00       2,011.00   
50%             3.00       35.00           1,997.00       2,012.00   
75%             5.00      230.00           2,004.00       2,013.00   
max            80.00   30,500.00           2,013.00       2,013.00   

       waterpoint_age  
count       57,565.00  
mean            15.04  
std             10.26  
min              0.00  
25%              7.00  
50%             14.00  
75%             17.00  
max             53.00

In [32]:
test_values_processed.describe()

id  amount_tsh  gps_height  longitude  latitude  region_code  \
count 14,850.00   14,850.00   14,850.00  14,850.00 14,850.00    14,850.00   
mean  37,161.97      322.83      655.15      34.06     -5.68        15.14   
std   21,359.36    2,510.97      691.26       6.59      2.94        17.19   
min       10.00        0.00      -57.00       0.00    -11.56         1.00   
25%   18,727.00        0.00        0.00      33.07     -8.44         5.00   
50%   37,361.50        0.00      344.00      34.90     -5.05        12.00   
75%   55,799.75       25.00    1,308.00      37.20     -3.32        17.00   
max   74,249.00  200,000.00    2,777.00      40.33     -0.00        99.00   

       district_code  population  construction_year  
count      14,850.00   14,850.00          14,850.00  
mean            5.63      184.11           1,997.06  
std             9.67      469.50               9.89  
min             0.00        0.00           1,960.00  
25%             2.00        0.00           1,996.00  
50%             3.00       20.00           1,997.00  
75%             5.00      220.00           2,004.00  
max            80.00   11,469.00           2,013.00